### DISCUSSION TITLE
关于用worker代理实现cloudflare单域名优选的想法

### DISCUSSION THREAD

**Discussion by kafuneri** [xingpingcn/enhanced-FaaS-in-China#11] [Ideas]
托管在cf的域名，子域名改NS记录托管到`dnspod`再指向优选IP的方法，对于`Pages`和`Workers路由`是可行的，但保护服务器源IP需要用两个域名来`Saas回源`；而对于`Zero Trus`t域名和`R2`储存桶自定义域名用`Saas回源`的方法也不能正常访问，但是有万能CF代理大法：

```js
export default {
  async fetch(request, env) {
    let url = new URL(request.url);
    if (url.pathname.startsWith('/')) {
      url.hostname = 'tunnels.domain.com'//要代理的域名
      let new_request = new Request(url, request);
      return fetch(new_request);
    }
    return env.ASSETS.fetch(request);
  },
};
```
`_worker.js`部署到`Pages`再绑定自定义域名，这样就能将自定义域名指向优选IP，经测试可以正常访问用`Zero Trust`的网站和`R2`储存桶，并且测速网站测出来的效果更好，但实际体验上并没有变快(反而感觉加载更慢了)，这样做是不是过了两遍CF节点所以起到了反效果?求大佬解答这样做有没有用

**Comment by xingpingcn** [comment]
cf worker用来反代的话好像过不了cf自己的盾（笑），会很卡。你的cf worker需要代理ip，但是这个我不懂。

有条件还是申请一个免费的二级域名来saas回源，如果这个免费域名被墙了应该问题也不大

**Reply by kafuneri** [reply]
好吧，感谢大佬回复。免费域名倒是有挺多，主要是用'CF隧道'内网穿透出来的网址用`Saas回源`的方法不能访问，直接404。服务器用`saas回源`的效果很好

**Comment by cloudeye2** [comment]
CF不允许套娃，即：

对于托管在cf上的worker或pages站点
1. 不允许用worker反代
2. 不允许做Saas的源站

**Reply by xingpingcn** [reply]
pages通过修改dns的方式来加速并没有使用到worker反代，也没有将其设为saas的源

**Reply by cloudeye2** [reply]
同意。
我意思是kafuneri提到的worker代理，是不能访问托管在cf上的站点的。（会引起CF网络的某种流量黑洞或者死循环？）
就像你推测的一样，如果非要那样做，需要一个外部代理。
